### DISCUSSION TITLE
Kafka pvcs out of sync in Argo

### DISCUSSION THREAD

**Discussion by popdaniel942** [strimzi/strimzi-kafka-operator#7205] [Q&A] [answered]
Starting with operator 0.29.0 all the kafka pvcs appear as out of sync in Argo, as if they need to get deleted. Trying to force sync or something still doesn't work. The pvcs get created but Argo says they need to be deleted. This didn't happen to previous versions of the operator. Is this normal behaviour? Did the creation of pvcs change?

**Comment by scholzj** [comment, marked as answer]
0.30.0 changed the default from StatefulSets to StrimziPodSets (there are IIRC other discussions on this on Slack and here on GitHub). But there was no change in 0.29.

**Reply by scholzj** [reply]
PS: I don't really use Argo, so I do not know how it tracks the PVCs. The PVCs by default do not have any ownerReference to avoid their automatic deletion (which is the Kubernetes default behavior). he `deleteClaim: true` flag in the storage configuration will set the owner reference since you request their deletion. So any changes to this value can IMHO affect the tracking in Argo.

**Reply by popdaniel942** [reply]
Yeah, most probably this is the problem, the fact that StatefulSets are no longer used. I think I have to set Argo to ignore pvcs.

**Reply by AydinChavez** [reply]
_I think I have to set Argo to ignore pvcs._
@popdaniel942 How did you manage this?

like this maybe?

```
   kafka:
      template:
        persistentVolumeClaim:
          metadata:
            annotations:
              argocd.argoproj.io/sync-options: Prune=false
```

**Reply by popdaniel942** [reply]
You can't tell Argo to ignore pvcs unfortuantely. If you have problems with Argo seeing the pvcs as out of sync in argo, you can either use a featureGate (-UseStrimziPodSet) in the Strimzi operator and it will revert to uisng StatefulSets instead of StrimziPodSets and the pvcs will no longer be out of sync. Or there is a setting for when you deploy argo https://argo-cd.readthedocs.io/en/latest/user-guide/resource_tracking/ . You can use  application.resourceTrackingMethod: annotation or  application.resourceTrackingMethod: annotation+label and argo will know better what resources actually need to be managed by Argo. I use this one and argo no longer sees the kafka pvcs

**Reply by popdaniel942** [reply]
This same thing was talked about here:
https://github.com/strimzi/strimzi-kafka-operator/issues/7093

**Reply by AydinChavez** [reply]
I am using following annotation for ignoring the pvcs created by strimzi in argocd sync state now (along with prune=false annotation): 

https://argo-cd.readthedocs.io/en/stable/user-guide/compare-options/

**Reply by scholzj** [reply]
Please keep in mind that the feature gate will not be there forever. So while short term it can help, it is not a long term solution. So if you use Argo, it would be good if you can figure out how does it track the PVCs with StatefulSets and if the same mechanism can be used for StrimziPodSets / Pods or if this is completely custom mechanism which works for StatefulSets only. Because for StatefulSets it's IMHO not tracked through ownerReference either.

**Comment by popdaniel942** [comment]
In my opinion I think the best solution is to use argo's annotation+label tracking, it will know much better what resources it actually needs to track

**Reply by scholzj** [reply]
Ok ... but ... what exactly does that mean for someone without any Argo experience? Is some additional support needed for it? Or does it work today?

Strimzi already sets the `app.kubernetes.io/instance` label to the name of the Kafka CR. This should be set on all resources including the PVCs. However, these labels are very badly defined, so it is never really clear what should the value be and everyone seems expect some different value or way how it is propagated.

**Reply by AydinChavez** [reply]
From my personal point of view a mechanism (via argocd specific annotations - as described in comment https://github.com/orgs/strimzi/discussions/7205#discussioncomment-3704150) exists in order to tell argocd to not consider the strimzi pvcs for deletion and also to ignore those in the overall sync state. For my usecase there is no further todo necessary in strimzi for this topic (if I didn’t overlook something).

**Comment by popdaniel942** [comment]
It's something available in the latest versions of argo, I am using it, it already works. There is nothing to do on kafka side, it's just a configuration when deploying argo.It will add an argo annotation instead of label or both( with annotation+label) to all resources that need to be managed by argo. And this way argo will know better what resources to manage, it will know if some resources were created by someone else. And I already tried this with kafka and it won't see it's pvcs so they can't get out of sync.

**Reply by ilja-maslov** [reply]
> It's something available in the latest versions of argo, I am using it, it already works. There is nothing to do on kafka side, it's just a configuration when deploying argo.It will add an argo annotation instead of label or both( with annotation+label) to all resources that need to be managed by argo. And this way argo will know better what resources to manage, it will know if some resources were created by someone else. And I already tried this with kafka and it won't see it's pvcs so they can't get out of sync.

The documentation https://argo-cd.readthedocs.io/en/latest/user-guide/resource_tracking/#additional-tracking-methods-via-an-annotation suggests that `application.resourceTrackingMethod` set to `annotation+label` will result in a specific annotation `argocd.argoproj.io/tracking-id` to be used to track the resources, but this annotation is not present in the Strimzi resources.

I think this explains the behavior you see, i.e. PVCs disappear from the Application screen.  I was considering using the annotation, but only if it would be possible to annotate every Strimzi object with it when deploying with a Helm chart.  AFAIK, this functionality is not available yet.  The idea is that `annotation+label` configuration, in conjunction with the ability to annotate every object (especially PVCs created with the SPS) would give us proper tracking in Argo of all the app objects and pruning could be then turned on and we'll see and can inspect these PVCs in the Argo UI :)

**Reply by scholzj** [reply]
> but this annotation is not present in the Strimzi resources

So, how can we make it _present_? This is not our annotation and we do not know what to set it to, or? Or can it be set to just anything and if the values match on all resources Argo will somehow track it? Or does it have some specific format where we say in the annotation that _this PVC is owned by a Pod `my-cluster-kafka-0`_ for example?

> I was considering using the annotation, but only if it would be possible to annotate every Strimzi object with it when deploying with a Helm chart. AFAIK, this functionality is not available yet. The idea is that annotation+label configuration, in conjunction with the ability to annotate every object (especially PVCs created with the SPS) would give us proper tracking in Argo of all the app objects and pruning could be then turned on and we'll see and can inspect these PVCs in the Argo UI :)

I think you can set annotations for all resources we create in the custom resource. You cannot set the annotation once and have it mirrored everywhere. Most labels set on the custom resource are mirrored to all resources. But for annotations this does not happen because there are many cases where you don't want to mirror the annotations (there are some for labels as well).

**Reply by ilja-maslov** [reply]
OK, I had some time today to play with the `tracking-id` Argo annotation.

Here is somewhat outdated [key discussion](https://github.com/argoproj/argo-cd/issues/8683) on how the references work and what exactly happens when they are not self-referenced.

TLDR; tracking-id value is expected to be either `my-argo-app-name:my-api/my-kind:my-namespace/my-name` or  `my-argo-app-name:other-api/other-kind:other-namespace/other-name`, i.e. the first part refers to the Argo application name, the other 2 parts form the resource reference, which could be either self-referenced or could reference another object.  When the resource under consideration is annotated referencing another resource, it will be represented in Argo UI at the top level, irrespective where in the hierarchy the referenced resource is.  So, if we reference the Argo application or Kafka cluster or a StrimziPod set, provided this resource is considered to be managed by Argo, it will also result in the annotated resource to be tracked by Argo too, but will be always shown in the UI, something along these lines:
<img width="670" alt="image" src="https://user-images.githubusercontent.com/78861727/191864744-2cab1ac2-31eb-42a2-9b83-2ed817eb2089.png">

Not everything is completely clean, but it is a workable solution, here is what it looks like atm.

1. We have `application.instanceLabelKey: argocd.argoproj.io/instance` in the argocd-cm to avoid the conflict between the default label Argo tracks (`app.kubernetes.io/instance`) which could be updated by various external helm charts or operators and the one used by Argo to track the resources belonging to the app.  This is a recommended Argo practice and the root cause why StrimziPod Set created PVCs don't seem to be associated with the Argo app
2. To enable annotation-based tracking in Argo CD, we need to add `application.resourceTrackingMethod: annotation+label` to the argocd-cm, this will switch the primary tracking to the annotations, yet still preserves the `ownerReference` that Argo uses implicitly to track the dependent resources
3. We annotate Argo app that manages the Strimzi cluster objects with a self-references tracking-id, because, by default, it would take the name from the parent App of apps pattern application:
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  annotations:
    argocd.argoproj.io/tracking-id: dev-strimzi-flex:argoproj.io/Application:argocd/dev-strimzi-flex
  name: dev-strimzi-flex
  namespace: argocd
...
```
4. We add annotations that reference the app to the Kafka CR, we could also reference the Kafka or even corresponding StrimziPodSet object, but the UI will still show the lines going from your Argo app directly to the PVCs, so it seems that there is very little value at the time to use direct dependency reference to track the PVCs
```
Kafka.spec.kafka.template:
      persistentVolumeClaim:
        metadata:
          annotations:
            argocd.argoproj.io/tracking-id: "dev-strimzi-flex:argoproj.io/Application:argocd/dev-strimzi-flex"
...
Kafka.spec.zookeeper.template:
      persistentVolumeClaim:
        metadata:
          annotations:
            argocd.argoproj.io/tracking-id: "dev-strimzi-flex:argoproj.io/Application:argocd/dev-strimzi-flex"
```
5. Argo app is green and synced, no orphan pruning is happening, the only gotcha is that when I click on one of the PVCs in the Argo UI, the detailed representation is not ideal (Unknown status and it shows the Diff between the desired and actual states, I'd really prefer Argo to show no diff and no STATUS, but it is a minor issue
<img width="635" alt="image" src="https://user-images.githubusercontent.com/78861727/191867646-52b988d0-e0fc-4392-9c0f-f24876b33f1b.png">

This is the best configuration I could have come up with so far.

One last thing, we run Argo 2.4.12, the argo issue link above suggests that the feature may be only working properly in 2.4.9+, but I haven't verified it.

**Reply by scholzj** [reply]
> we could also reference the Kafka or even corresponding StrimziPodSet object

How would the annotation look like in this case? I guess it would it be something like this for example:

```
argocd.argoproj.io/tracking-id: "my-argo-app-name:core.strimzi.io/StrimziPodSet:my-namespace/my-cluster-kafka"
```

So it will still have the `my-argo-app-name` part, right? I wondered if there is some way for us to set this automatically, but we don't know the Argo Application name, so it would not be possible I guess.

**Reply by ilja-maslov** [reply]
> > we could also reference the Kafka or even corresponding StrimziPodSet object
> 
> How would the annotation look like in this case? I guess it would it be something like this for example:
> 
> ```
> argocd.argoproj.io/tracking-id: "my-argo-app-name:core.strimzi.io/StrimziPodSet:my-namespace/my-cluster-kafka"
> ```
Yes, the last part refers to the api/kind referenced in the middle part, so`my-cluster-kafka` or `my-cluster-zookeeper` depending on  the SPS you'd like to reference. There are 2 modes for the reference, either you self-reference or not, when you are not self-referencing, but referencing any resource in the cluster (preserving the first part of the tracking-id), the sync becomes green and sync status Unknown for the PVC (see previous post).

Now, I have tried to self-reference for just one of the PVCs (PersistentClaimStorageOverride doesn't allow to override the annotations per broker, anyway) and I do not think I like what I see (non-self-referenced PVCs look much better, I think they must match on some prefix, maybe?):
<img width="652" alt="image" src="https://user-images.githubusercontent.com/78861727/192026904-414e8412-ec87-47c3-8224-8b9be2a5df3f.png">
<img width="560" alt="image" src="https://user-images.githubusercontent.com/78861727/192027149-2233da6a-4907-4ed1-bdf9-a79012fff07f.png">
It gets back to the same representation I had before turning on the annotation tracking, weird, probably still some bugs there.

I have also added `argocd.argoproj.io/sync-options: Prune=false` to PVC annotations to be able to prune the other resources automatically.

I've got a feeling from that feature discussion link from the previous post that Argo team prefers to display the non-self referenced resources at the top level in the UI, i.e. direct arrow from the Argo app to the resource.  Maybe, I'll still point it to the corresponding SPS so that when they change their mind, we won't need to re-point :)

 Update: ended up with these annotations, similar for the zookeeper:
```
      persistentVolumeClaim:
        metadata:
          annotations:
            argocd.argoproj.io/sync-options: Prune=false
            argocd.argoproj.io/tracking-id: "dev-strimzi-flex:core.strimzi.io/StrimziPodSet:dev-kafka/dev-cluster-kafka"
```

**Reply by AydinChavez** [reply]
What about following approach (the one I use):

Setting in the Argo config (as you do):

	application.instanceLabelKey: argocd.argoproj.io/instance

Along with setting both following annotations in the PVC template from Strimzi:

	argocd.argoproj.io/compare-options: IgnoreExtraneous
	argocd.argoproj.io/sync-options: Prune=False

It should help to get rid of the faulty sync/diff results in ArgoCD.

**Reply by ilja-maslov** [reply]
The way I understood it (haven't tried), IgnoreExtraneous will just paint the Argo app green instead of yellow, but PVCs will stay out of sync.  I believe that our ideal state is that we are not ignoring the out of sync errors, but rather teaching Argo to track these PVCs correctly as dependencies.  Doesn't seem like there is a clean way to fully achieve this yet, unless one sets `deleteClaim: true` but this is highly undesirable for any production deployment.

**Comment by dbaarorapr** [comment]
@ilja-ag do you have a answer to above issues,we have same problems now that we need to remove the feature flag that allows us to use stateful sets as we need to upgrade to 0.36.1 strimzi and argo shows the out of sync (requires pruning) for PVCS and CRB?

**Comment by domenicbove** [comment]
Looks like Argo recommends this:
```
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: kafka
spec:
  kafka:
    template:
      persistentVolumeClaim:
        metadata:
          annotations:
            argocd.argoproj.io/compare-options: IgnoreExtraneous
            argocd.argoproj.io/sync-options: Prune=false
  zookeeper:
    template:
      persistentVolumeClaim:
        metadata:
          annotations:
            argocd.argoproj.io/compare-options: IgnoreExtraneous
            argocd.argoproj.io/sync-options: Prune=false
```

In the Argo UI it shows the Application is Synced, even though the PVCs are not synced. I think that's why also want you to set prune false to be safe!

**Comment by VijayPatil872** [comment]
Hi @scholzj,
We have deployed kafka with kRaft, with strimzi operator version 0.45.0. We are facing similar kind of issue related to PVCs. The ArgoCD is out of sync and PVCs are going in to progressing giving status pending deletion. All the kafka broker and controllers PVC appear as out of sync in ArgoCD, as if they need to get deleted. The PVCs get created but ArgoCD says they need to be deleted. We tried some workaround as stated above also we tried some workaround related to ignoreDifferences in ArgoCD application manifests but still issue persists. Is there any workable solution for this issue we can have??

**Reply by scholzj** [reply]
I have no idea. I do not know a thing about Argo and I do not use it.
